### PR TITLE
[HOLD] Improve error handling

### LINKED ISSUE #137: Error handling does not currently comply to the invoker specification
The invoker specification details error conditions [here](https://github.com/projectriff/invoker-specification/blob/3ddb2ac1fe521f01dad3f0b6c57a9f9ab13a6f89/streaming.md#stream-completion-and-error-conditions).

What is currently missing:

 - `When an invoker receives the error stream completion signal on its input stream, it MUST propagate that error termination to all of the function input streams.`
 - `When a function emits an error completion signal on one of its several output streams, an invoker MUST immediately forward that signal back to the main output stream.`
 - when a function throws _when invoked_ (as opposed to when receiving data [see this point below]), the error should be propagated to the gRPC output stream (this is part of the spec yet but should be)

What is currently implemented and should be removed:

- the streaming pipeline currently terminates if any of the input unmarshaller errors (see [this](https://github.com/projectriff/node-function-invoker/blob/2883a0cf551be1f58d1605484ebf285d1b47b268/lib/streaming-pipeline.js#L162) and [this](https://github.com/projectriff/node-function-invoker/blob/2883a0cf551be1f58d1605484ebf285d1b47b268/lib/streaming-pipeline.js#L45))

What is currently unclear:

 - how to handle functions doing something like (with fictional map operator):
```js
inputStream
    .pipe(map(() => throw new Error('yolo')))
    .pipe(outputStream)
```

### PR DESCRIPTION
**Pull request by fbiville**
Fixes #137.

### CODE REVIEW

**Comment by fbiville** [comment]
@scothis @ericbottard PR rebased and completely revised

**Comment by fbiville** [comment]
2 things related to unmarshaller error handling that bug me still:
- [this](https://github.com/projectriff/node-function-invoker/blob/d1843ef17a5bb6d6871732d0cedfa96084774ed2/lib/streaming-pipeline.js#L217) does not _feel_ right. Is any error from the unmarshaller a valid cause to cancel the gRPC invocation?
- [this](https://github.com/projectriff/node-function-invoker/blob/d1843ef17a5bb6d6871732d0cedfa96084774ed2/lib/input-unmarshaller.js#L58) is not right. `callback` calls should not be handled in a try-catch. However, simply getting rid of the try-catch here leads to [this test failure](https://github.com/projectriff/node-function-invoker/blob/d1843ef17a5bb6d6871732d0cedfa96084774ed2/spec/streaming-pipeline.errors.spec.js#L296).

Ideas welcome, @scothis @ericbottard

**Comment by fbiville** [comment]
Another problem: now that logical outputs forward errors to the main output (i.e. the gRPC Duplex stream), the latter emits an error.
However, this Duplex stream is also the main input, which means that the same error will be interpreted as a source error as well and will be forwarded to all the unmarshallers.

@scothis do you know if there is a way to distinguish errors from the Readable and Writable side of a Duplex stream?

**Comment by scothis** [comment]
> do you know if there is a way to distinguish errors from the Readable and Writable side of a Duplex stream?

You don't. Logically Readable and Writable are interfaces. Duplex combines both interfaces into a single interface. The implementing object has an error, not the interface. Granted, JavaScript blurs this because there's no such thing as an interface in the language.

If it's important to split the Readable and Writable interfaces to exist on different objects, don't use a Duplex stream. But I'd first step back and consider if your view of the problem space is correct. The first question I'd ask is if your abusing the error. For a grpc stream, I'd expect an error to be a fault in the socket, and not an application concern where inputs and outputs have any semantic distinction.

caveat: I'm not an expert on node streams.

**Comment by scothis** [comment]
> this does not feel right. Is any error from the unmarshaller a valid cause to cancel the gRPC invocation?

What's the suggested course of action to recover from the error? Either we need to fail, or give some way for the user to recover (which could be messy). Casually ignoring the error is no good.

> this is not right. callback calls should not be handled in a try-catch. However, simply getting rid of the try-catch here leads to this test failure.

Calls to `callback()` are inconsistently in a try block. Perhaps move the start of the final try block to the top of the function so that all calls to catch are handled.

**Comment by fbiville** [comment]
> don't use a Duplex stream. 
No choice here, that's how `node-grpc` is implemented. Main input is the Readable side, main output the Writable side.

**Comment by fbiville** [comment]
I think this PR introduces a more logical way of handling errors and is worth merging as-is.
Now rebased, PTAL.

File: lib/streaming-pipeline.js [hunk 1]
```diff
@@ -225,10 +238,14 @@ module.exports = class StreamingPipeline extends Writable {
         for (let i = 0; i < outputCount; i++) {
             const marshaller = new OutputMarshaller(i, outputContentTypes[i]);
             marshaller.pipe(this._destinationStream, {end: false});
+            marshaller.on('error', (err) => {
```
Comment (scothis) [inline lib/streaming-pipeline.js]: why is this needed? pipe forwards errors

File: lib/streaming-pipeline.js [hunk 2]
```diff
@@ -40,9 +40,18 @@ module.exports = class StreamingPipeline extends Writable {
         this._functionOutputs = [];
         this._finishedOutputs = 0;
         this._streamMetadata = {};
-        this.on('finish', () => {
-            logger('Ending input streams');
-            this._endInputs();
+        this.on('pipe', (src) => {
```
Comment (scothis) [inline lib/streaming-pipeline.js]: the end and error events should be propagated by pipe.

File: lib/streaming-pipeline.js [hunk 3]
```diff
@@ -118,6 +127,10 @@ module.exports = class StreamingPipeline extends Writable {
         this._functionInputs.forEach((fi) => fi.emit('end'));
     }
 
+    _errorInputs(err) {
+        this._functionInputs.forEach((fi) => fi.emit('error', err));
```
Comment (scothis) [inline lib/streaming-pipeline.js]: I think this is backwards. What I would expect to see is something that listen to each input stream passed into the function for unmarshaling errors, since we can't assume the user function will handle that correctly. If any of the input streams err, pass that error back via the grpc stream.

File: lib/streaming-pipeline.js (hunk 2 above)
Comment (fbiville) [reply lib/streaming-pipeline.js]: I don't see that in the [documentation](https://nodejs.org/docs/latest-v12.x/api/stream.html#stream_readable_pipe_destination_options).
The `end` _function_ of the Writable is called when the Readable ends.
Nothing about error propagation.

File: lib/streaming-pipeline.js (hunk 3 above)
Comment (fbiville) [reply lib/streaming-pipeline.js]: I'm following the [spec](https://github.com/projectriff/invoker-specification/blob/master/streaming.md#stream-completion-and-error-conditions) here:
```
When an invoker receives the error stream completion signal on its input stream, it MUST propagate that error termination to all of the function input streams.
```

File: lib/streaming-pipeline.js (hunk 2 above)
Comment (scothis) [reply lib/streaming-pipeline.js]: you're right, I misread [this snippet](https://github.com/nodejs/node/blob/4c746a6cfda980c1cd0de6246781c0083d9e416c/lib/_stream_readable.js#L737-L755). It's watching the destination stream on pipe for errors, not the source.
